core/templates/dev/head/services/contextual/UrlServiceSpec.js

```diff
@@ -67,22 +58,41 @@ describe('Url Service', function() {
         UrlService.getQueryFieldValuesAsList('field2')).toEqual(expectedList);
     });
 
-  it('should correctly get parameter list based on key value', function() {
-    var expectedList = ['parent1', 'parent2'];
-    window.href = 'http://' + pathname + queryString;
-    expect(
-      UrlService.getQueryFieldValuesAsList('parent')).toEqual(expectedList);
-    window.href = 'http://' + pathname;
-    expect(UrlService.getQueryFieldValuesAsList('parent')).toEqual([]);
-  });
+  it('should correctly encode and encode special characters in URI',
+    function() {
+      window.href = 'http://' + pathname;
+      var expectedObject = {
+        field1: '?value=1'
+      };
+      window.href = UrlService.addField(window.href, 'field1', '?value=1');
```

Comment:
Should be `var url =`, right? External callers to the service should not be changing stuff the service depends on. You'll want to get the url and then expect(url).toEqual(something).

Reply:
In fact, in this whole test, don't modify window.href at all. The addField() method is supposed to be non-state-changing.

Have separate tests that are focused on getQueryFieldValuesAsList() and get UrlParams(), and put those with the other tests for those methods. In those tests it is fine to set the mockLocation, then call the function and verify what it returns, because those methods depend on the mockLocation. But make sure to keep the setting of the mockLocation separate from actually conducting the test. E.g.:

```
    mockLocation = ...;
    expect(UrlService.someMethod()).toEqual(...);

    mockLocation = ...;
    expect(UrlService.someMethod()).toEqual(...);
```

Reply:
Ok, will do that. So, addField() and UrlSevice.getUrlParams() or getQueryFieldValuesAsList() shouldn't be in same test together.

Reply:
Well, yeah. I considered it, but it's probably cleaner to do it function by function.

Reply:
Ok, done